src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/HalfConverter.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal sealed class HalfConverter : JsonPrimitiveConverter<Half>
+    {
+        private const int MaxFormatLength = 16;
+        private const int MaxEscapedFormatLength = MaxFormatLength * JsonConstants.MaxExpansionFactorWhileEscaping;
+
+        public HalfConverter()
+        {
+            IsInternalConverterForNumberType = true;
+        }
+
+        public override Half Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            if (reader.TokenType != JsonTokenType.Number)
+            {
+                ThrowHelper.ThrowInvalidOperationException_ExpectedNumber(reader.TokenType);
+            }
+
+            return ReadCore(ref reader);
+        }
+
+        public override void Write(Utf8JsonWriter writer, Half value, JsonSerializerOptions options)
+        {
+            WriteCore(writer, value);
+        }
+
+        private static Half ReadCore(ref Utf8JsonReader reader)
+        {
+            Half result;
+
+            if (reader.ValueLength > MaxEscapedFormatLength)
+            {
+                ThrowHelper.ThrowFormatException(NumericType.Half);
+            }
+
+            Span<byte> buffer = stackalloc byte[MaxFormatLength];
+            int written = reader.CopyString(buffer);
+            buffer = buffer.Slice(0, written);
+
+            if (reader.TokenType is JsonTokenType.String or JsonTokenType.PropertyName)
+            {
+                if (JsonReaderHelper.TryGetFloatingPointConstant(buffer, out result))
```

Comment:
Isn't it the case that we only try to read floating point literals if `JsonNumberHandling.AllowNamedFloatingPointLiterals` is enabled? This appears to be the default deserialization path.

Reply:
No, if you look at `SingleConverter`:
https://github.com/dotnet/runtime/blob/1380a0b327a24a924ef3f63f144558c742da0a3c/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/SingleConverter.cs#L37-L49

When `AllowReadingFromString` is specified, we call GetSingleWithQuotes() which checks both floating point literals and numbers.

I agree it sounds counterintuitive because JsonNumberHandling is a Flags enum, but that's how it is.